### ISSUE TITLE
DDF CORE: Unexpected behavior for missing InstanceType

### ISSUE THREAD

**Issue by eljanssens** [cdisc-org/cdisc-rules-engine#1546]
This issue has been noted by Mike Rippin from Capgemini (see attached email)

[DDF CORE_Unhandled CORE Errors - UPDATE.pdf](https://github.com/user-attachments/files/24719255/DDF.CORE_Unhandled.CORE.Errors.-.UPDATE.pdf)

_Any entity defined that is missing the "InstanceType" attribute causes CORE to generate and unhandled exception.
It will be much better if the CORE engine catches such issues and reports them as part of the Conformance Report, and preferably identifies the element hierarchy that led to the error so a user knows where to find it in the source file._

Summary of the issue:

- The check is performed correctly in the CORE engine.
- However, the way that it is picked up and reported is depending on a certain attribute in the data (the instanceType). That helps in identifying where the problem is located in the USDM.

If that attribute instanceType is missing then nothing is reported now, while it is a problem that it is missing and it is intended to be reported.

@OGarcia11: can this be assigned to @gerrycampion or @SFJohnson24 ?

**Comment by pendingintent** [comment]
@gerrycampion , @SFJohnson24 are you able to take this issue?

**Comment by SFJohnson24** [comment]
@pendingintent @eljanssens I was outside the DDF project scope so I am not the most familiar with USDM and how it operates within engine.  I dont want to put more on @gerrycampion plate but I believe he would be much better suited than me to address this issue expeditiously.  I am happy to take it but I will need a bit of time getting up to speed on USDM

**Comment by pendingintent** [comment]
Latest Suggestion:
 
_It looks like the problem is that – although the schema validation does identify the problems with the referenceSubstance – the error messages are not associated with a “dataset”/entity (because instanceType is missing) so they’re not reported.  I think the schema checking logic should be updated to assign the “dataset”/entity from the class/entity derived from the schema if it’s not available from the data._

**Comment by pendingintent** [comment]
Link to the "dirty" file: https://github.com/cdisc-org/cdisc-jsonata-rules/blob/main/testdata/CDISC_Pilot_Study_v4_Dirty.json
Link to the "fixed" file: https://github.com/cdisc-org/cdisc-jsonata-rules/blob/main/testdata/CDISC_Pilot_Study_v4_FIXED.json

**Comment by pendingintent** [comment]
./core validate -s USDM -v 4.0 -d ../core_files/dirty

[CORE-Report-2026-01-26T09-15-51.xlsx](https://github.com/user-attachments/files/24861357/CORE-Report-2026-01-26T09-15-51.xlsx)

**Comment by pendingintent** [comment]
./core validate -s USDM -v 4.0 -d ../core_files/fixed

[CORE-Report-2026-01-26T09-17-15.xlsx](https://github.com/user-attachments/files/24861376/CORE-Report-2026-01-26T09-17-15.xlsx)

**Comment by pendingintent** [comment]
I am running the executable on Mac since I do not have access to a Windows instance.

I am not able to reproduce the exception when executing CORE on the DIRTY file when it is hte only file in the directory but when I copy both the `dirty` file and the `fixed` file into the same directory, an exception is thrown:

dmoreland@Mac core % ./core validate -s USDM -v 4.0 -d ../core_files/dirty
  [------------------------------------]    0%
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "multiprocessing/pool.py", line 125, in worker
  File "scripts/run_validation.py", line 98, in validate_single_rule
  File "cdisc_rules_engine/rules_engine.py", line 148, in validate_single_rule
AttributeError: 'NoneType' object has no attribute 'get'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "core.py", line 875, in <module>
  File "click/core.py", line 1157, in __call__
  File "click/core.py", line 1078, in main
  File "click/core.py", line 1688, in invoke
  File "click/core.py", line 1434, in invoke
  File "click/core.py", line 783, in invoke
  File "click/decorators.py", line 33, in new_func
  File "core.py", line 466, in validate
  File "scripts/run_validation.py", line 192, in run_validation
  File "cdisc_rules_engine/utilities/progress_displayers.py", line 62, in _bar_progress_displayer
  File "multiprocessing/pool.py", line 873, in next
AttributeError: 'NoneType' object has no attribute 'get'
[PYI-24478:ERROR] Failed to execute script 'core' due to unhandled exception!

**Comment by pendingintent** [comment]
**multiprocessing/pool.py:121-171**

<img width="715" height="923" alt="Image" src="https://github.com/user-attachments/assets/f09baca1-bac2-4596-b286-9f6cf9f7595b" />



**core.py:875**
cli.add_command(list_rule_sets)

**cdisc_rules_engine/utilities/progress_displayers.py:47-65**

<img width="487" height="347" alt="Image" src="https://github.com/user-attachments/assets/bdc3c9c2-af32-4696-bb63-22ced7573233" />

**Comment by pendingintent** [comment]
The error is occurring at [rules_engine.py:148](vscode-webview://0rt28ih25rog93fbf3sv6fe3hf2du9tjahi4bta63h958cihhibb/cdisc_rules_engine/rules_engine.py#L148) where it's trying to call .get() on rule["domains"], but rule["domains"] is None

code checks if "domains" exists as a key in the rule dictionary, but it doesn't verify that rule["domains"] isn't None. When the rule has "domains": None, the check passes but then line 148 fails trying to call .get() on None

The code is checking if `domains` key exists but does not check if not `None`.

Edit to make safe:
./cdisc_rules_engine/rules_engine.py:147-150
if dataset_metadata.unsplit_name in results and "domains" in rule:
                    **include_split = rule.get("domains", {}).get("include_split_datasets", False)**
                    if not include_split:
                        continue  # handling split datasets

**Comment by ASL-rmarshall** [comment]
@pendingintent In your [comment above](https://github.com/cdisc-org/cdisc-rules-engine/issues/1546#issuecomment-3799874912), I noticed that you're using the `-d` argument to run the CLI.  For validation of single JSON files that contains multiple "datasets", we usually use the `-dp` argument and point directly to the file (not the directory/folder):
```
./core validate -s USDM -v 4.0 -dp ../core_files/dirty/CDISC_Pilot_Study_v4_Dirty.json
```

**Comment by pendingintent** [comment]
I have used @ASL-rmarshall suggestion and tested only a single file:
dmoreland@Mac core % ./core validate -s USDM -v 4.0 -dp ../core_files/1546/dirty/CDISC_Pilot_Study_v4_Dirty.json 
  [████████████████████████████████████]  100%
[DEBUG 2026-01-26 10:33:38,137 - excel_report.py:90] - Report written to: CORE-Report-2026-01-26T10-33-29.xlsx
Output: CORE-Report-2026-01-26T10-33-29

dmoreland@Mac core % ./core validate -s USDM -v 4.0 -dp ../core_files/1546/dirty/CDISC_Pilot_Study_v4_FIXED.json 
  [████████████████████████████████████]  100%
[DEBUG 2026-01-26 10:34:45,101 - excel_report.py:90] - Report written to: CORE-Report-2026-01-26T10-34-36.xlsx
Output: CORE-Report-2026-01-26T10-34-36

The file(s) does not report any issues when tested alone.  If using the `-d` flag and testing more than one JSON file in the directory, the exception is generated.  If best practice is to use `-dp`, we ned to make this clear.

**Comment by pendingintent** [comment]
Some additional feedback:

I think there are two aspects to this issue:
1. Lack of an instanceType value may cause CORE to throw an unhandled error.  I agree that the code should be updated so that the exception is not thrown.

2. The lack of instanceType prevents the correct reporting of schema validation issues (if these issues had been reported correctly, Mike would have been able to identify and correct the errors that resulted in the exception being thrown).  The code that creates a dataset of schema validation issues (cdisc_rules_engine\dataset_builders\json_schema_check_dataset_builder.py) needs to be updated to assign the “dataset” value to be the schema class name when instanceType is not available in the data (currently at L113).

**Comment by pendingintent** [comment]
Line 113: errlist["dataset"].append(errctx.get("instanceType", "") if errctx else "")

Change line 113 to extract the schema class name from the error's schema $ref when instanceType is not available

i.e., get the instanceType from errctx, but if instanceType is not available, use the schema name (simliar to block starting at 136) and only fall back to empty string ("") if schema name is also unavailable.

**Comment by pendingintent** [comment]
I have tested with a dirty file I created on my own and can see that when an entity instanceType is removed, the exception is thrown:

dmoreland@Mac core % ./core validate -s USDM -v 4.0 -dp /Users/dmoreland/Downloads/pilot_LLZT_protocol_DIRTY.json 
Traceback (most recent call last):
  File "cdisc_rules_engine/services/data_readers/json_reader.py", line 12, in from_file
  File "json/__init__.py", line 293, in load
  File "json/__init__.py", line 346, in loads
  File "json/decoder.py", line 338, in decode
  File "json/decoder.py", line 354, in raw_decode
json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 52 column 11 (char 1894)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "core.py", line 875, in <module>
  File "click/core.py", line 1157, in __call__
  File "click/core.py", line 1078, in main
  File "click/core.py", line 1688, in invoke
  File "click/core.py", line 1434, in invoke
  File "click/core.py", line 783, in invoke
  File "click/decorators.py", line 33, in new_func
  File "core.py", line 466, in validate
  File "scripts/run_validation.py", line 151, in run_validation
  File "cdisc_rules_engine/services/data_services/data_service_factory.py", line 59, in get_data_service
  File "cdisc_rules_engine/services/data_services/usdm_data_service.py", line 486, in is_valid_data
  File "cdisc_rules_engine/services/data_readers/json_reader.py", line 15, in from_file
cdisc_rules_engine.exceptions.custom_exceptions.InvalidJSONFormat: 
  Error reading JSON from: /Users/dmoreland/Downloads/pilot_LLZT_protocol_DIRTY.json
  JSONDecodeError: Expecting property name enclosed in double quotes: line 52 column 11 (char 1894)
[PYI-66665:ERROR] Failed to execute script 'core' due to unhandled exception!

I have remvoed the `instanceType` property from `geographicScopes` at line 51:

"instanceType": "GovernanceDate"

[pilot_LLZT_protocol_DIRTY.json](https://github.com/user-attachments/files/24917507/pilot_LLZT_protocol_DIRTY.json)

**Comment by ASL-rmarshall** [comment]
@pendingintent You're getting that specific error because you didn't remove the comma at the end of line 50, so the JSON is invalid.

**Comment by gerrycampion** [comment]
Because engine is using `instanceType` for processing and because we might not have a rule to check for `instanceType`, this should probably be a check during the load/preprocessing steps with better error messages.

**Comment by pendingintent** [comment]
Update from user:



Thanks for this message.  I am including ***** who has replaced me on the DDF team (though she is a Capgemini colleague and I will continue to support her).

I do think this bug is important to fix as it prevents the SDR continuing to execute the CORE rules, and as initially reported it took a lot of time to debug - this may well affect other users as well.  On this point, I think it will be true of any unhandled exception, so if there is at least a top-level catch all and some ability to track whatever the current file line number, or parent entity was being processed, would help future debugging and source correction.

Thanks,

**Comment by ASL-rmarshall** [comment]
> Because engine is using `instanceType` for processing and because we might not have a rule to check for `instanceType`, this should probably be a check during the load/preprocessing steps with better error messages.

There is a rule to check for `instanceType` - `instanceType` is defined as a required attribute in the USDM schema and DDF00125 is designed to report when required attributes are missing.  This is the rule that is not reporting issues correctly - the schema violation is identified by the CORE process that builds the initial dataset for the JSON Schema Check rule type, but "dataset" is currently not populated when `instanceType` is missing.  This means that the schema violations are not included when DDF00125 reports for each of the datasets identified for the study.  Once the JSON Schema Check process has been updated to retrieve the class name from the schema when `instanceType` is missing, this will solve some of the issues - but only if there is at least one (other) instance of the relevant class that does have `instanceType`.  If there isn't another instance with `instanceType`, there won't be a dataset created for the class name - the dataset created by the USDM data service will be named according to the fall-back processing (I think it gets the name of the containing attribute) - so the issue still won't be reported.

Given that `instanceType` is so important I agree that its absence should probably be reported as an issue during the load/preprocessing steps.  If we do this, we can probably get rid of the original look-up mechanism that uses resources\schema\USDM.yaml to assign dataset/class name when `instanceType` is missing. Alternatively, you could replace the static USDM.yaml file with one that's created dynamically from the schema.
